### PR TITLE
Be My Ardent-tine

### PR DESCRIPTION
**Pull request by allenh1**
This is a work in progress still! I'm not super pleased with the quality of this patch.

But, in its current state, this _does_ add ROS 2 support, which is pretty neat.

### CODE REVIEW

**Comment by allenh1** [comment]
Next up is to add support for things like python (for using distutils instead of CMake), as, currently, ROS2 ebuilds fail due to a lack of `CMakeLists.txt` file.

Also, there's complaints about a missing opensplice dependency, so nothing is buildable past the type support.

File: superflore/generators/ebuild/overlay_instance.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ def commit_changes(self, distro):
             'lunar': 'regenerate ros-lunar, ',
             'indigo': 'regenerate ros-indigo, ',
             'kinetic': 'regenerate ros-kinetic, ',
+            'ardent': 'regenerage ros2-ardent, ',
```
Comment (tfoote) [inline superflore/generators/ebuild/overlay_instance.py]: typo 'regenerage'

**Comment by nuclearsandwich** [comment]
> Next up is to add support for things like python (for using distutils instead of CMake), as, currently, ROS2 ebuilds fail due to a lack of CMakeLists.txt file.

It would be a big change, and maybe not worth making at the moment, but adding ebuild generation to bloom and then using it in superflore might make sense long term.

Since 0.6.0 bloom uses different templates for generating different package build types. Even if it doesn't make sense to try and incorporate bloom now, you can still see how the debian packages are constructed. https://github.com/ros-infrastructure/bloom/tree/master/bloom/generators/debian/templates

**Comment by allenh1** [comment]
Ok, I'm a little upset about these random failures on Travis. Under "pr", the CI passes for pypy3, and for "push" it fails only for python 3.5. And for string orderings? I don't get it.

Must be some `LC_LANG` setting or something that's disrupting the sort function.

File: superflore/generators/ebuild/ebuild.py [hunk 2]
```diff
@@ -225,19 +225,28 @@ def get_ebuild_text(self, distributor, license_text):
         # CMAKE_BUILD_TYPE
         if self.name == "catkin":
             ret += "BUILD_BINARY=\"0\"\n"
+
         ret += "ROS_DISTRO=\"{0}\"\n".format(self.distro)
         ret += "ROS_PREFIX=\"opt/ros/${ROS_DISTRO}\"\n"
-
+        # TODO(allenh1): This is gross. No.
```
Comment (nuclearsandwich) [inline superflore/generators/ebuild/ebuild.py]: I'm regularly guilty of this as well, but instead of saying "This is gross." A more useful XXX or TODO comment might give an example of when the code will fail

Comment (allenh1) [reply superflore/generators/ebuild/ebuild.py]: It's just there so I fix it before this gets merged (I will not merge that into the codebase -- that's just awful).

Comment (allenh1) [reply superflore/generators/ebuild/ebuild.py]: Actually, on second thought, I really don't think this is as awful as I initially thought it was.

The more robust solution is to create an `eclass` for this, but that's not really in the spirit of things if it's just for unpacking things. So I think I'll just leave a note that says "if you need to update this, then do so. Otherwise, this is sufficient."

**Comment by allenh1** [comment]
I think this just about does it for this PR. I ended up adding in a template `ament-python.eclass`. I think there should be a follow up with some tests for the new logic introduced here, but I don't want to add that to this PR (it's already pretty bloated).

tl;dr The remaining fixes will be done to the eclasses, so this can be merged.

Next step there is getting @nuclearsandwich's `ros_workspace` to build. Currently, it's [not](https://gist.github.com/allenh1/48e80007fc07efafccd9561bdaded8c9) (likely due to a multilib mapping issue).

File: superflore/generators/ebuild/run.py [hunk 3]
```diff
@@ -31,12 +32,16 @@
...
 from superflore.utils import load_pr
 from superflore.utils import ok
+from superflore.utils import ros2_distros
 from superflore.utils import save_pr
 from superflore.utils import url_to_repo_org
 from superflore.utils import warn
 
-# Modify if a new distro is added
-active_distros = ['indigo', 'kinetic', 'lunar']
+# TODO(allenh1): It would be super nice make this a configuration option.
```
Comment (tfoote) [inline superflore/generators/ebuild/run.py]: :+1:  `ROSDISTRO_INDEX_URL`

Just use `rosdistro.get_index_url()` https://github.com/ros-infrastructure/rosdistro/blob/master/src/rosdistro/__init__.py#L68

Comment (allenh1) [reply superflore/generators/ebuild/run.py]: I figured there was going to be something like that. I'll look into this in a follow up. Thanks.
